src/components/action-popover/action-popover.component.js

```diff
@@ -0,0 +1,251 @@
+import React, {
+  useState, useCallback, useEffect, useRef
+} from 'react';
+import PropTypes from 'prop-types';
+import I18n from 'i18n-js';
+import {
+  Menu, MenuItemFactory, MenuButton, ButtonIcon, MenuItemIcon, MenuItemDivider
+} from './action-popover.style';
+import Events from '../../utils/helpers/events';
+import createGuid from '../../utils/helpers/guid';
+import OptionsHelper from '../../utils/helpers/options-helper';
+
+const ActionPopover = ({ children, id }) => {
+  const [isOpen, setOpen] = useState(false);
+  const [focusIndex, setFocusIndex] = useState(0);
+  const [childrenWithRef, setChildrenWithRef] = useState();
+  const [items, setItems] = useState();
+  const [guid] = useState(createGuid());
+  const button = useRef();
+
+  // Keyboard commands implemented as reccomended by WAI-ARIA best practices
+  // https://www.w3.org/TR/wai-aria-practices/examples/menu-button/menu-button-actions.html
+
+  const onButtonKeyDown = useCallback(((e) => {
+    if (Events.isSpaceKey(e) || Events.isDownKey(e) || Events.isEnterKey(e)) {
+      e.preventDefault();
+      setOpen(!isOpen);
+      if (isOpen) {
```

Comment:
Shouldn't these conditions be the other way around?

Reply:
So this is checking the current state, before the next render. If it's currently open and we've just pressed Enter we want it to focus the button. Enter is the only event that propagates from the menu item so this code wouldn't be called if the user pressed space/down.

If it's currently closed, and we're about to render the open state we want to set the focus on the first element.

Reply:
Ah yep, that makes sense. The actions are the opposite way round in the `onButtonClick` callback (lines 115-125) - does that mean those need changing?

Reply:
I see what you mean, I'll refactor that now so they're consistent.

Reply:
Resolved in https://github.com/Sage/carbon/pull/2308/commits/758bc43ea1dd974568b7269ec6332496b358b936